File: backend/app/schema/application_schema.py
from pydantic import BaseModel
from typing import List, Optional
from datetime import datetime

class EmailMetadata(BaseModel):
    sender: str
    subject: str
    body: str
    received_at: datetime

class Attachment(BaseModel):
    type: str
    file_path: str
    uploaded_at: datetime

class Merchant(BaseModel):
    legal_name: str
    dba_name: str
    federal_tax_id: str
    address: str
    industry: str
    annual_revenue: float

class Owner(BaseModel):
    name: str
    ssn: str
    address: str
    date_of_birth: datetime
    ownership_percentage: float

class FundingDetails(BaseModel):
    amount_requested: float
    use_of_funds: str

class Application(BaseModel):
    id: str
    status: str
    created_at: datetime
    updated_at: datetime
    email_metadata: EmailMetadata
    attachments: List[Attachment]
    merchant: Merchant
    owners: List[Owner]
    funding_details: FundingDetails

# HUMAN ASSISTANCE NEEDED
# Consider adding validation rules for fields like SSN, federal_tax_id, and email format.
# Also, consider adding additional optional fields that might be useful for the application process.

File: frontend/src/schema/applicationSchema.ts
import { z } from 'zod';

interface EmailMetadata {
  sender: string;
  subject: string;
  body: string;
  receivedAt: Date;
}

interface Attachment {
  type: string;
  filePath: string;
  uploadedAt: Date;
}

interface Merchant {
  legalName: string;
  dbaName: string;
  federalTaxId: string;
  address: string;
  industry: string;
  annualRevenue: number;
}

interface Owner {
  name: string;
  ssn: string;
  address: string;
  dateOfBirth: Date;
  ownershipPercentage: number;
}

interface FundingDetails {
  amountRequested: number;
  useOfFunds: string;
}

interface Application {
  id: string;
  status: string;
  createdAt: Date;
  updatedAt: Date;
  emailMetadata: EmailMetadata;
  attachments: Attachment[];
  merchant: Merchant;
  owners: Owner[];
  fundingDetails: FundingDetails;
}

const ApplicationSchema = z.object({
  id: z.string(),
  status: z.string(),
  createdAt: z.date(),
  updatedAt: z.date(),
  emailMetadata: z.object({
    sender: z.string(),
    subject: z.string(),
    body: z.string(),
    receivedAt: z.date(),
  }),
  attachments: z.array(z.object({
    type: z.string(),
    filePath: z.string(),
    uploadedAt: z.date(),
  })),
  merchant: z.object({
    legalName: z.string(),
    dbaName: z.string(),
    federalTaxId: z.string(),
    address: z.string(),
    industry: z.string(),
    annualRevenue: z.number(),
  }),
  owners: z.array(z.object({
    name: z.string(),
    ssn: z.string(),
    address: z.string(),
    dateOfBirth: z.date(),
    ownershipPercentage: z.number(),
  })),
  fundingDetails: z.object({
    amountRequested: z.number(),
    useOfFunds: z.string(),
  }),
});

export type { Application, EmailMetadata, Attachment, Merchant, Owner, FundingDetails };
export { ApplicationSchema };